### PR TITLE
8293488: Add EOR3 backend rule for aarch64 SHA3 extension

### PR DESCRIPTION
**Pull request by Bhavana-Kilambi**
Arm ISA v8.2A and v9.0A include SHA3 feature extensions and one of those SHA3 instructions - "eor3" performs an exclusive OR of three vectors. This is helpful in applications that have multiple, consecutive "eor" operations which can be reduced by clubbing them into fewer operations using the "eor3" instruction. For example -
```
eor a, a, b
eor a, a, c
```
can be optimized to single instruction - `eor3 a, b, c`

This patch adds backend rules for Neon and SVE2 "eor3" instructions and a micro benchmark to assess the performance gains with this patch. Following are the results of the included micro benchmark on a 128-bit aarch64 machine that supports Neon, SVE2 and SHA3 features -

```
Benchmark               gain
TestEor3.test1Int       10.87%
TestEor3.test1Long      8.84%
TestEor3.test2Int       21.68%
TestEor3.test2Long      21.04%
```

The numbers shown are performance gains with using Neon eor3 instruction over the master branch that uses multiple "eor" instructions instead. Similar gains can be observed with the SVE2 "eor3" version as well since the "eor3" instruction is unpredicated and the machine under test uses a maximum vector width of 128 bits which makes the SVE2 code generation very similar to the one with Neon.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293488](https://bugs.openjdk.org/browse/JDK-8293488): Add EOR3 backend rule for aarch64 SHA3 extension

### Reviewers
 * [Hao Sun](https://openjdk.org/census#haosun) (@shqking - Author) ⚠️ Review applies to [6df4f014](https://git.openjdk.org/jdk/pull/10407/files/6df4f014c648fdb916437aeebea1f8385d7cdb32)
 * [Ningsheng Jian](https://openjdk.org/census#njian) (@nsjian - Committer) ⚠️ Review applies to [a0aa8cdc](https://git.openjdk.org/jdk/pull/10407/files/a0aa8cdca243e16e93ece10db16498a4eee034f1)
 * [Eric Liu](https://openjdk.org/census#eliu) (@theRealELiu - Committer) ⚠️ Review applies to [7f413360](https://git.openjdk.org/jdk/pull/10407/files/7f4133602f3064322fb48328dbfc7261b3df816e)
 * [Andrey Turbanov](https://openjdk.org/census#aturbanov) (@turbanoff - Committer) ⚠️ Review applies to [7f413360](https://git.openjdk.org/jdk/pull/10407/files/7f4133602f3064322fb48328dbfc7261b3df816e)
 * [Nick Gasson](https://openjdk.org/census#ngasson) (@nick-arm - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10407/head:pull/10407` \
`$ git checkout pull/10407`

Update a local copy of the PR: \
`$ git checkout pull/10407` \
`$ git pull https://git.openjdk.org/jdk pull/10407/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10407`

View PR using the GUI difftool: \
`$ git pr show -t 10407`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10407.diff">https://git.openjdk.org/jdk/pull/10407.diff</a>

</details>

### CODE REVIEW

**Comment by Bhavana-Kilambi** [comment]
Could anyone please take a look at this PR and give their feedback ? Thank you ..

File: test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    // For the IR verification of "veor3_sve" backend node in the below tests,
+    // "svesha3" CPU feature is supposed to be checked as it is an optional
+    // feature in SVE2 but it is enabled in all the currently available SVE2
+    // hardware. Hence that check is omitted.
+
+    // Test for eor3 Neon and SVE2 instruction for integers
+    @Test
+    @IR(counts = {"veor3_neon", "> 0"}, applyIf = {"MaxVectorSize", "16"})
+    @IR(counts = {"veor3_sve", "> 0"}, applyIfAnd = {"UseSVE", "2", "MaxVectorSize", "> 16"})
```
Comment (shqking) [inline test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java:87]:
```suggestion
    @IR(counts = {"veor3_sve", "> 0"}, applyIfAnd = {"UseSVE", "2", "MaxVectorSize", "> 16"}, applyIfCPUFeature = {"svesha3", "true"})
```

After this PR(https://github.com/openjdk/jdk/pull/10402), `applyIf` and `applyIfCPUFeature` are evaluated as a logical conjunction. We can check CPU features and VM options at the same time now. Of course, the comment at line 79 should be removed.

File: test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.Random;
+
+import jdk.test.lib.Utils;
+
+/*
+ * @test
+ * @bug 8293488
+ * @summary Test EOR3 Neon/SVE2 instruction for aarch64 SHA3 extension
+ * @library /test/lib /
+ * @requires os.arch == "aarch64" & vm.cpu.features ~=".*sha3.*"
```
Comment (shqking) [inline test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java:38]:
```suggestion
 * @requires os.arch == "aarch64" & vm.cpu.features ~= ".*sha3.*"
```
nit: [style] it's better to have one extra space.

Comment (Bhavana-Kilambi) [reply test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java]: @shqking  Thank you for the comments. I have made the suggested changes.

**Review by shqking — APPROVED** [review]
LGTM (I'm not a Reviewer).

**Review by nsjian — APPROVED** [review]
lgtm

File: test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+ * @summary Test EOR3 Neon/SVE2 instruction for aarch64 SHA3 extension
+ * @library /test/lib /
+ * @requires os.arch == "aarch64" & vm.cpu.features ~= ".*sha3.*"
+ * @run driver compiler.vectorization.TestEor3AArch64
+ */
+
+public class TestEor3AArch64 {
+
+    private final static int LENGTH = 2048;
+    private final static Random RD = Utils.getRandomInstance();
```
Comment (turbanoff) [inline test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java:45]: let's use blessed modifiers order
```suggestion
    private static final Random RD = Utils.getRandomInstance();
```

File: test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+ * @bug 8293488
+ * @summary Test EOR3 Neon/SVE2 instruction for aarch64 SHA3 extension
+ * @library /test/lib /
+ * @requires os.arch == "aarch64" & vm.cpu.features ~= ".*sha3.*"
+ * @run driver compiler.vectorization.TestEor3AArch64
+ */
+
+public class TestEor3AArch64 {
+
+    private final static int LENGTH = 2048;
```
Comment (turbanoff) [inline test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java:44]:
```suggestion
    private static final int LENGTH = 2048;
```

Comment (Bhavana-Kilambi) [reply test/hotspot/jtreg/compiler/vectorization/TestEor3AArch64.java]: Hello, thank you for your feedback. I have made the suggested changes and uploaded a new patch. Please review ..

**Review by theRealELiu — APPROVED** [review]
LGTM.

**Comment by Bhavana-Kilambi** [comment]
The new patch removes the svesha3 feature check for eor3 instruction. Eor3 instruction is part of the SHA3 feature but it is present by default in SVE2 and is not part of the SVESHA3 feature.

**Comment by Bhavana-Kilambi** [comment]
@turbanoff  Hello, I have made the changes you've suggested plus some more changes regarding the feature detection for svesha3 in the latest patch. Could you please take a look? Thank you in advance ..

**Review by turbanoff — APPROVED** [review]

**Review by nsjian — APPROVED** [review]

**Review by theRealELiu — APPROVED** [review]

**Comment by Bhavana-Kilambi** [comment]
Hello, I have resolved merge conflicts and have uploaded the latest patch here. Please review. I messed up with the re-request review option but I now understand how it works (I thought I could re-request from everyone but realized that the previous selection is removed if i select another reviewer). Apologies if any inconvenience caused.

**Review by nsjian — APPROVED** [review]

File: test/hotspot/gtest/aarch64/aarch64-asmtest.py [hunk 5]
```diff
@@ -1040,7 +1040,10 @@ def cstr(self):
                         width +
                         [str(self.reg[i]) for i in range(1, self.numRegs)]))
     def astr(self):
-        formatStr = "%s%s" + ''.join([", %s" for i in range(1, self.numRegs)])
+        if self._name == "eor3":
```
[the remark this answers is not in the document]
Comment (Bhavana-Kilambi) [reply test/hotspot/gtest/aarch64/aarch64-asmtest.py]: Thank you for the suggestion. I made the suggested changes in the latest patch. Please review.

**Review by nick-arm — APPROVED** [review]

**Pull request closed without merging**